autoload/lsp/ui/vim.vim

```diff
@@ -347,25 +410,141 @@ function! s:apply_workspace_edits(workspace_edits) abort
+    let l:cmd = l:cmd . " | execute 'keepjumps normal! " . l:escaped_sub_cmd . "'"
+
+    call lsp#log('s:build_cmd', l:cmd)
+
+    return l:cmd
+endfunction
+
+function! s:generate_sub_cmd(text_edit) abort
+    if a:text_edit['range']['start']['line'] == a:text_edit['range']['end']['line'] &&
+        \ a:text_edit['range']['start']['character'] == a:text_edit['range']['end']['character']
+        return s:generate_sub_cmd_insert(a:text_edit)
+    else
+        return s:generate_sub_cmd_replace(a:text_edit)
+    endif
+endfunction
+
+function! s:generate_sub_cmd_insert(text_edit) abort
+    let l:start_line = a:text_edit['range']['start']['line']
+    let l:start_character = a:text_edit['range']['start']['character']
+    call lsp#log('REMOVE', a:text_edit['newText'])
+    let l:new_text = s:parse(a:text_edit['newText'])
+    call lsp#log('REMOVE', l:new_text)
+
+    let l:sub_cmd = s:generate_move_cmd(l:start_line, l:start_character)
+
+    if len(l:new_text) == 0
+        let l:sub_cmd .= 'x'
+    else
+        if a:text_edit['range']['start']['character'] >= len(getline(a:text_edit['range']['start']['line']))
+            let l:sub_cmd .= 'a'
+        else
+            let l:sub_cmd .= 'i'
+        endif
+    endif
+
+    let l:sub_cmd .= printf('%s', l:new_text)
+
+    return l:sub_cmd
+endfunction
+
+function! s:generate_sub_cmd_replace(text_edit) abort
+    let l:start_line = a:text_edit['range']['start']['line']
+    let l:start_character = a:text_edit['range']['start']['character']
+    let l:end_line = a:text_edit['range']['end']['line']
+    let l:end_character = a:text_edit['range']['end']['character'] - 1  " -1 since the last character is excluded
+    let l:new_text = a:text_edit['newText']
+
+    let l:sub_cmd = s:generate_move_cmd(l:start_line, l:start_character) " move to the first position
+
+    if len(l:new_text) == 0
+        let l:sub_cmd .= 'x'
+    else
+        let l:sub_cmd .= 'v'
+    endif
+
+    let l:sub_cmd .= s:generate_move_cmd(l:end_line, l:end_character) " move to the last position
+    let l:sub_cmd .= printf('c%s', l:new_text) " change text
+
+    return l:sub_cmd
+endfunction
+
+function! s:generate_move_cmd(line_pos, character_pos) abort
+    let l:result = printf('%dG0', a:line_pos) " move the line and set to the cursor at the beginning
+    if a:character_pos > 0
+        let l:result .= printf('%dl', a:character_pos) " Move right until the character
+    endif
+    return l:result
+endfunction
+
+function! s:parse(text) abort
+    " https://stackoverflow.com/questions/71417/why-is-r-a-newline-for-vim
+    return substitute(a:text, '\(\n$\|\r\n\)', '\r', 'g')
+endfunction
+
+" https://microsoft.github.io/language-server-protocol/specification#text-documents
+" Position in a text document expressed as zero-based line and zero-based
+" character offset.
+function! s:parse_range(range) abort
+   let a:range['start']['line'] =  a:range['start']['line'] + 1
```

Comment:
I did the change but I am not sure if I got it right

Reply:
actually need to use `deepcopy` since you are modifying nested objects

---
In reply to: [208430551](https://github.com/prabirshrestha/vim-lsp/pull/162#discussion_r208430551) [](ancestors = 208430551)